File: HTML/ArabicLexerBase.py
import unicodedata
from antlr4 import Lexer, InputStream  # <-- لاحظ أننا أضفنا InputStream هنا


class ArabicLexerBase(Lexer):
    """
    Base class that strips Arabic diacritics (Tashkeel, U+064B–U+065F
    and U+0670) from the input *before* ANTLR tokenises it.
    """

    _DIACRITICS = set(range(0x064B, 0x0660)) | {0x0670}

    def __init__(self, input_stream, output=None):
        # 1. نقرأ النص الخام بالكامل من الصندوق القديم
        raw_text = input_stream.getText(0, input_stream.size - 1)
        
        # 2. ننظف النص من التشكيل
        cleaned = self._strip_tashkeel(raw_text)
        
        # 3. نصنع صندوقاً جديداً (Stream جديد) نضع فيه النص النظيف
        new_stream = InputStream(cleaned)
        
        # 4. نسلم الصندوق الجديد للبرنامج ليكمل عمله بدون مشاكل
        super().__init__(new_stream, output)

    @classmethod
    def _strip_tashkeel(cls, text: str) -> str:
        """Return *text* with all Arabic diacritic characters removed."""
        return "".join(ch for ch in text if ord(ch) not in cls._DIACRITICS)

File: Grammar/ArabicLexerBase.java
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.Lexer;
import org.antlr.v4.runtime.misc.Interval;

/**
 * Base class that strips Arabic diacritics (Tashkeel) from the input 
 * *before* ANTLR tokenises it, while preserving accurate line/column tracking.
 */
public abstract class ArabicLexerBase extends Lexer {

    // Removed the PrintStream constructor as ANTLR4 Java Lexer doesn't support it
    public ArabicLexerBase(CharStream input) {
        super(wrapStream(input));
    }

    private static CharStream wrapStream(CharStream input) {
        if (input instanceof ArabicCharStream) {
            return input;
        }
        
        // Extract the original text safely from the provided stream
        input.seek(0);
        String rawText = input.getText(Interval.of(0, input.size() - 1));
        
        // Wrap the original text in our custom stream
        return new ArabicCharStream(rawText);
    }

    /**
     * Overrides the default Lexer line tracking to return the 
     * precomputed original line number from the un-diacriticized text.
     */
    @Override
    public int getLine() {
        if (_input instanceof ArabicCharStream) {
            return ((ArabicCharStream) _input).getOriginalLine();
        }
        return super.getLine();
    }

    /**
     * Overrides the default Lexer column tracking to return the 
     * precomputed original column number from the un-diacriticized text.
     */
    @Override
    public int getCharPositionInLine() {
        if (_input instanceof ArabicCharStream) {
            return ((ArabicCharStream) _input).getOriginalColumn();
        }
        return super.getCharPositionInLine();
    }
}